components/brave_wallet/browser/zcash/zcash_rpc.cc

```diff
@@ -0,0 +1,207 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_rpc.h"
+
+#include "base/big_endian.h"
+#include "base/functional/bind.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "net/base/load_flags.h"
+#include "net/http/http_request_headers.h"
+#include "net/http/http_status_code.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/mojom/url_loader.mojom.h"
+
+namespace brave_wallet::zcash_rpc {
+
+namespace {
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("zcash_rpc", R"(
+      semantics {
+        sender: "ZCash RPC"
+        description:
+          "This service is used to communicate with ZCash Lightwalletd nodes "
+          "on behalf of the user interacting with the native Brave wallet."
+        trigger:
+          "Triggered by uses of the native Brave wallet."
+        data:
+          "ZCash JSON RPC response bodies."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on chrome://flags."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+bool UrlPathEndsWithSlash(const GURL& base_url) {
+  auto path_piece = base_url.path_piece();
+  return !path_piece.empty() && path_piece.back() == '/';
+}
+
+const GURL MakeGetAddressUtxosURL(const GURL& base_url) {
+  if (!base_url.is_valid()) {
+    return GURL();
+  }
+  if (!UrlPathEndsWithSlash(base_url)) {
+    return GURL();
+  }
+
+  GURL::Replacements replacements;
+  std::string path =
+      base::StrCat({base_url.path(),
+                    "cash.z.wallet.sdk.rpc.CompactTxStreamer/GetAddressUtxos"});
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+// Prefixes provided serialized protobuf with compression byte and 4 bytes of
+// message size. See
+// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+std::string GetPrefixedProtobuf(const std::string& serialized_proto) {
+  char compression = 0;  // 0 means no compression
+  char buff[4];          // big-endian 4 bytes of message size
+  base::WriteBigEndian<uint32_t>(buff, serialized_proto.size());
+  std::string result;
+  result.append(&compression, 1);
+  result.append(buff, 4);
+  result.append(serialized_proto);
+  return result;
+}
+
+std::string MakeGetAddressUtxosURLParams(
+    const std::vector<std::string>& addresses) {
+  zcash::GetAddressUtxosRequest request;
+  for (const auto& address : addresses) {
+    request.add_addresses(address);
+  }
+  request.set_maxentries(1);
+  request.set_startheight(0);
+  return GetPrefixedProtobuf(request.SerializeAsString());
+}
+
```

Comment:
Is it ok from privacy perspective to send account's addresses in a batch address?

Reply:
Generally no, although if these are one time use addresses it's less concerning but if we can avoid it we should.

Reply:
Now it takes only 1 address